### ISSUE TITLE
SeqFeature.extract() fails when feature has negative offsets

### ISSUE THREAD

**Issue by kblin** [biopython/biopython#584]
On a GenBank file with negative numbers in the feature location line, Biopython will throw a warning like: `BiopythonParserWarning: Couldn't parse feature location: '-2..159'` and then continue. However, this will cause the downstream processing of the featuers to fail:
```
  File "Bio/SeqFeature.py", line 354, in extract
    return self.location.extract(parent_sequence)
AttributeError: 'NoneType' object has no attribute 'extract'
```

**Comment by kblin** [comment]
So the thing is that the genbank parser doesn't create a FeatureLocation object for negative offsets. Is this the intended behaviour? That is: is the correct fix to add an "if not self.location is None" check to SeqFeature.extract(), or do we instead fix the parser to always create a FeatureLocation?

**Comment by kblin** [comment]
It looks like we used to (incorrectly) parse this until bc733da09, but IMHO that patch is doing the right thing.
The code explicitly seems to give a "None" feature location on parsing issues, so I guess the correct way to handle this is to fix SeqFeature.extract()? What would be a good return on a "None" location extract()? As we seem to be unwilling to raise a LocationParserError in the parser, raising some sort of error in the extract doesn't feel right.

**Comment by peterjc** [comment]
Being strict, we could fail immediately during parsing on the invalid location, but that breaks a lot of use cases. Therefore we raise a ``BiopythonParserWarning`` allowing the user to process most of the file. Using ``None`` for the ``SeqFeature``'s ``.location`` seems reasonable to me here.

Rather than the ``AttributeError``, perhaps calling ``.extract(...)`` when the ``.location is None``, I suggest we should raise a ``ValueError`` with a more helpful message (i.e. this feature has no location)?

The only other behaviours I can think of is returning an empty sequence, but that seems risky to me, or having ``.extract(...)`` return ``None``?

**Comment by kblin** [comment]
Ok, I think there's two things we can do. I like the idea of the ValueError in extract. That way the error message isn't confusing.
And talking with a co-worker during lunch we had this idea of a "strict mode" and a "relaxed mode" during parsing, where "relaxed mode" would be what we do right now, and "strict mode" raising a ParserError everywhere we currently issue a warning.

**Comment by peterjc** [comment]
We have this strict/permissive mode in the PDB parser, but I think it is overly complex. To me the more elegant option is use the ``warnings`` library to treat ``BiopythonParserWarning`` as an error if you want to be strict.

Do you want to make a PR on the ``ValueError`` in the ``.extract(...)`` method?

**Comment by kblin** [comment]
Sure, working on this right now.

**Comment by kblin** [comment]
I guess this was fixed by #585.

**Comment by peterjc** [comment]
Yes, thanks - I should have close this.
